### PR TITLE
Add ability to find events by slide text & captions in search

### LINKED ISSUE #677: Make events findable by transcript (ideally jump to timestamp)
A really nice feature of [ETH's current video portal](https://video.ethz.ch/) is that you can search for parts of a transcript. It then shows you videos where your search term was said and also has a visualization when in the video it was said. If you click on the search result, you can directly jump to the timestamp when it was said.

image

We want that in Tobira as well! Several areas are involved here. First we need the transcription data with timestamps from Opencast. Then we need to put those into Meilisearch. And finally, our frontend needs to display those search results.

Blocked on:
- #191 
- #620

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
Fixes #677

## For testers

| [Test](https://pr1189.tobira.opencast.org/) |
| --- |

This PR does not contain any changes to the search page except adding this timeline. This is planned for later. This PR is already big enough.

Also note that the usefulness and the UX of this feature depends a lot on the available data! On our test instance, roughly 2500 events have OCR'ed slide texts, while only very few have subtitles. I will try to upload more videos soon to simply have more videos with subtitles available.  Subtitle timespans are usually shorter (in the order of seconds or 10s), while the timespans associated with slide text can have durations of many minutes.

### Questions/discussions

- Unlike the old video portal, this shows the actual text that was matched (with a small context). I find this cool, but it of course somewhat exposes how bad the OCR slide text and automatic subtitles sometimes are. 
- What do you think about the timeline design and how the matched text is highlighted?
- Report any query that leads to "internal server error" please. That should obviously never happen. 

### Search terms to get started

While testing myself I found a few good queries to get started. Of course, do try your own ones and also try prefixes of these to see how well it works. Also try multiple query words.
- `open`: big mixed bag
- `meilisearch`: finds two Tobira videos talking about Meilisearch (never mentioned in metadata)
- `tycho`: finds the "Tycho crater" in the NASA moon video subtitles AND text detection
- `crater`: lots of usages in the subtitles of the NASA moon video
- `pyroxene`: finds the mineral in NASA moon subtitles
- `elasticsearch`: lots of matches in Opencast-related videos
- `postgres`: showing some videos with "postgres" in title first (makes sense) and only then once that only mention postgres
- `videoportal`: obvious Tobira videos, but also one unrelated video screen-sharing the old ETH video portal briefly and one mentioning "videoportal" in its slides
- `feynman`: further down lots of videos just mentioning feynman 

## Technical info

This PR has these main parts:
- Add DB table `event_texts` (for storing all texts belonging to an event)
- Add `event_texts_queue` and process to automatically download text assets from Opencast (this is ran as part of the worker)
  - This was the most tricky part actually, in order to make it robust against random errors, OC or otherwise. To work well enough in most cases, without ever running into a super busy loop or something like that.
- Various helper sub commands to manage fetching assets
- VTT and MPEG7 parsers to parse the text assets
- Add texts to MeiliSearch in a special encoded form to optimize for Meili-search-performance while still allowing us to figure out the timespans of a match
- Make frontend use this data and show a timeline with matches for events

This can be mostly reviewed commit by commit. There are two times where I move a big chunk of code around that was added in a previous commit, but it should be fairly clear what and where. 

Performance is kind of important for this one, since we are dealing with potentially lots of data. So far it seems like Meili responds within 25ms in all cases I tested. That's fine, but still a big increase from before. We should make sure that we don't accidentally introduce some slowness. Though right now I also have no idea how we would optimize further....

Something I want to improve in a follow up PR: replace the busy polling in the "download assets" and "update search index" workers by `LISTEN/NOTIFY` events from Postgres. Right now, both default to 30s or sth, which means that adding an event has quite a round trip (sync + 30s + 30s) before its text assets are searchable. That can be vastly reduced. But again, this PR is already big enough.

### CODE REVIEW

File: backend/src/args.rs [hunk 1]
```diff
@@ -146,10 +146,10 @@ fn parse_color_choice(s: &str) -> Result<ColorChoice, &'static str> {
 
 impl Args {
     pub(crate) fn stdout_color(&self) -> ColorChoice {
-        if std::io::stdout().is_terminal() { self.color } else { ColorChoice::Never }
+        self.color
     }
 
     pub(crate) fn stderr_color(&self) -> ColorChoice {
-        if std::io::stderr().is_terminal() { self.color } else { ColorChoice::Never }
+        self.color
```
Comment (owi92) [inline backend/src/args.rs:153]: To be fair, I didn't change what this does in https://github.com/elan-ev/tobira/commit/acac4afb351614ce5491d545ae3558048f5117b3, so I can't tell you why.

**Review by owi92 — APPROVED** [review]
Looked through the code and tested a bunch but didn't find any obvious issues.
I'll do a final round of testing today and then this can be merged. My comments are of no real concern and definitely no blockers.

File: frontend/src/routes/Search.tsx [hunk 2]
```diff
@@ -817,4 +856,43 @@ export const handleNavigation = ((router: RouterControl, ref?: RefObject<HTMLInp
...
+
+        // UTF-8 chars have either the first bit 0 or the first two bits 1.
+        return v < 0x80 || v >= 0xC0;
+    };
+
+    const bytes = new TextEncoder().encode(s);
+    const decoder = new TextDecoder("utf-8");
+
+    // Round indices to avoid cutting into UTF8 chars. The loop only needs to
+    // execute 3 times as every 4 bytes there is always a char boundary.
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply frontend/src/routes/Search.tsx:903]: Strict rule: 

> UTF-8 is capable of encoding all 1,112,064 valid Unicode [code points](https://en.wikipedia.org/wiki/Code_point) using one to four one-[byte](https://en.wikipedia.org/wiki/Byte) (8-bit) code units.

So 4 bytes maximum, which means that if we are not already on a char boundary, after 3 bumps, we must be.

[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply frontend/src/routes/Search.tsx:903]: > text extraction and transcription won't ever output any special chars

That's not true, they definitely do. We should basically always assume that any unicode char can occur.

**Review by owi92 — APPROVED** [review]

**Comment by LukasKalbertodt** [comment]
@oas777 @dagraf Maybe the two of you want to take a short look at this before our Wednesday meeting. See the top comment for more information. But if you don't have the time till then, no worries, there will be plenty of more time to discuss search-related stuff before this is released.

**Comment by oas777** [comment]
First of all it's good to see this in action, thanks. Some initial observations:

- https://pr1189.tobira.opencast.org/~search?q=tyco tells me we're currently not distinguishing where the results come from, right?
- https://pr1189.tobira.opencast.org/~search?q=opencast tells me we're looking for pages, series, and videos, right? 
- In conjunction, I think we might have to make these distinctions clearer for users to understand the search results they are looking at (and their order in the sense of importance also).
- Also, with the number of results for https://pr1189.tobira.opencast.org/~search?q=internet we probably have to think about filters.
- https://pr1189.tobira.opencast.org/~search?q=schulte (blush) providing results for "schule" to me indicates search terms are too open.
-  I prefer a preview of the slides to a preview of the text extracted. The image also is an additional help to remember a certain part of the lecture you are looking for.
- Design: The timeline looks odd, mainly because highlighted segments hover over the timeline. I prefer having them "stringed" on the actual timeline.
- Font: The line with "Part of series" looks odd, even though it's probably supposed to be related to the line with the creator's name.

**Comment by LukasKalbertodt** [comment]
> https://pr1189.tobira.opencast.org/~search?q=tyco tells me we're currently not distinguishing where the results come from, right?

Do you mean "slide text" vs "captions"? Yes, currently both are treated as one thing. Is that different in your current portal?

> https://pr1189.tobira.opencast.org/~search?q=opencast tells me we're looking for pages, series, and videos, right?

Correct, which was like that already before. There will be some improvements there in an upcoming PR, like combining a series with the page listing only that series, as having these as two separate results is fairly useless.

> In conjunction, I think we might have to make these distinctions clearer for users to understand the search results they are looking at (and their order in the sense of importance also).

That is also something I'm planning to do in the upcoming PR. I am not sure if I will succeed with that, as it requires clever design, but yeah: my goal is that it's clear at a glance whether I'm looking at a video (should should be most results), a series or something else.

> Also, with the number of results for https://pr1189.tobira.opencast.org/~search?q=internet we probably have to think about filters.

Filters are of course planned already, and in fact some basic ones are already implemented. That feature is still hidden though, and will be reenabled with, you guessed it, my upcoming PR.

Apart from that, I would expect most users to just specify more query terms. I can't imagine a scenario where someone wants to find a video that they just remember had "internet" in it. And thanks to the clever ranking, users can just add a bunch of query words that they think are relevant, and the result containing most of these words will be shown first. Not to say we don't want filters -- we do -- but these search engines make filters *less* necessary as just adding more search terms usually works out.

> https://pr1189.tobira.opencast.org/~search?q=schulte (blush) providing results for "schule" to me indicates search terms are too open.

Mh I'm not sure I agree. That's typo tolerance in action. All videos by you (with an exact "schulte" match) are sorted before all other videos. So in my book that's exactly as it should be. And as last resort, you can always search by `"schulte"` (with quotes), which works exactly like in Google or most other search engines: looking for that term *exactly*.

> I prefer a preview of the slides to a preview of the text extracted. The image also is an additional help to remember a certain part of the lecture you are looking for.

Mh fair, the image seems useful. We don't always have an image though, especially for search results in captions, it might not be clear what to show. And: would you not show the extracted text at all then? I think it's useful.

> Design: The timeline looks odd, mainly because highlighted segments hover over the timeline. I prefer having them "stringed" on the actual timeline.

So more like the design in your current video portal?

> Font: The line with "Part of series" looks odd, even though it's probably supposed to be related to the line with the creator's name.

Not sure I understand.

**Comment by oas777** [comment]
> So more like the design in your current video portal?
> 
> > Font: The line with "Part of series" looks odd, even though it's probably supposed to be related to the line with the creator's name.
> 
> Not sure I understand.

Weird coincidence, but it's similar to what I [just said](https://github.com/elan-ev/tobira/issues/1196#issuecomment-2206112365) for Paella: 

grafik

looks like five different fonts for five different text elements.

**Comment by oas777** [comment]
For reference, here's how Kaltura organises search results in the [UZH video portal](https://uzh.mediaspace.cast.switch.ch/esearch/search?keyword=einstein).

- Their search seems "too open" as well, providing results for "Stein" might be ok, but "Einstellungen" and "Kleinstaaten" blurs results.
- I like the fact that you can filter this a) by source and b) date - they call it relevance, but I think "date ascending/descending" and "semester" would be perfect.
- Clear distinction between "Channels" and videos, which we also need, especially if we add "pages".

**Comment by dagraf** [comment]
> For reference, here's how Kaltura organises search results in the [UZH video portal](https://uzh.mediaspace.cast.switch.ch/esearch/search?keyword=einstein).
> 
>     * Their search seems "too open" as well, providing results for "Stein" might be ok, but "Einstellungen" and "Kleinstaaten" blurs results; not sure why "Die" is also listed.
I agree.
> 
>     * I like the fact that you can filter this a) by source and b) date - they call it relevance, but I think "date ascending/descending" and "semester" would be perfect.
Me too.
> 
>     * Clear distinction between "Channels" and videos, which we also need, especially if we add "pages".
I agree.

Additionally:
- I like the fact that the expressions or words that are responsible for the video to show up in the search result being highlighted.
-

**Comment by LukasKalbertodt** [comment]
Okay, that took a while but I think this PR is now ready. 

@oas777 @dagraf please have another look. The timeline is now closer to the design of the ETH's current video portal ("stringed") and preview images for that video section are shown (IF they are available!!). Further, results from captions and slides are displayed slightly differently. This also means that we can now distinguish between these, meaning that in the future we could add options to just search through slides, not captions, for example. 

Let me know what you think about the functionality and the design of the timeline showing text matches. Note again: UI stuff not related to the timeline is not part of this PR and a follow up PR will polish the search page design.

**Comment by oas777** [comment]
Thanks, Lukas and Ole. My first impression is positive and I'll try to restrict my comments to the timeline:
- For slides results, skip the text extract, avoiding gibberish like "Schuleetal 13032024 7" and distinguishing results more clearly from those in audio/transcript.
- For transcript results, maybe add quotation marks? "…Entwicklung scheint das Zauberwort der Zukun 211 sein und… "
- I'm not sure, but are you adding the slide preview for (some) transcript results? I saw some without slide, some with, the latter being confusing (search "Lars" and his talk "New Integration in BigBlueButton").
- Maybe distinguish the two in color in the timeline?
- Or are you doing this already? Search "video" and see David's talk: 
grafik
- The size of the white canvas in which slide and transcript results are presented seems to depend on the lenght of the text found which makes it uneasy to look at (first result for "video", Matthias' talk).
- Same goes for the fact that some results are above, some are below the timeline.

**Comment by LukasKalbertodt** [comment]
> * For slides results, skip the text extract, avoiding gibberish like "Schuleetal 13032024 7" and distinguishing results more clearly from those in audio/transcript.

It's usually not possible to see the actual matched text in the slide due to the low resolution. And the matched part wouldn't be highlighted as it is now. I think the actual matched text can help a lot with giving context. But of course, showing gibberish is not useful either. So I will make it configurable at least.

> * For transcript results, maybe add quotation marks? "…Entwicklung scheint das Zauberwort der Zukun 211 sein und… "

That's a good idea. Unfortunately, there is hidden complexity. There are many different kinds of quotation marks and using the right one depends on the language of the text. „German”, “English”, »German alternative and Danish«, « French », «Swiss French and German?» ... just to name a few of languages/cultures familiar to us. [There are way more](https://en.wikipedia.org/wiki/Quotation_mark#Summary_table). And the problem is knowing the language of the search result. We can check the language of the video (which is already annoying but ok), but that might not be set or not match the text in question. It might be possible to only use one style of quotation mark by having them far away from the text and stylized somehow, but still... I would probably just leave as is.

> * I'm not sure, but are you adding the slide preview for (some) transcript results? I saw some without slide, some with, the latter being confusing (search "Lars" and his talk "New Integration in BigBlueButton").

The slide preview is shown for both kinds of results (caption or slide text), but only for videos which offer slide previews at all. My reasoning was: even if its a caption match, seeing the slide might still ease finding the spot in the video you are looking for. Or do you think showing the preview for caption results implies too strongly that the result is from the slide? 

> * Maybe distinguish the two in color in the timeline?
> * Or are you doing this already? Search "video" and see David's talk:

Yes, this is already done. It's just a different brightness of the color, not a different hue. But we don't really have a different-hue-color we could use (except the "danger color", but that's not fitting).

> * The size of the white canvas in which slide and transcript results are presented seems to depend on the lenght of the text found which makes it uneasy to look at (first result for "video", Matthias' talk).

Yeah true, fair point, will try to fix that.

> * Same goes for the fact that some results are above, some are below the timeline.

They are always shown above, except if there is not enough space above. I think that's a useful behavior.

**Comment by LukasKalbertodt** [comment]
> > * The size of the white canvas in which slide and transcript results are presented seems to depend on the lenght of the text found which makes it uneasy to look at (first result for "video", Matthias' talk).
> 
> Yeah true, fair point, will try to fix that.

I spoke too quickly: how would you even fix this? The problematic case is when the matched text is long. I think there are only these options, none of which is optimal:

- Cut off the matched text, only showing a very tiny section around the match: I think the length of the shown text is already very short and reducing the context around the match isn't great.
- Let the text wrap into more than two lines: also not sure how useful that is.
- Make the image larger: blowing up these low resolution images isn't great, and having such a big tooltip isn't great either.

**Comment by dagraf** [comment]
> > * For slides results, skip the text extract, avoiding gibberish like "Schuleetal 13032024 7" and distinguishing results more clearly from those in audio/transcript.
> 
> It's usually not possible to see the actual matched text in the slide due to the low resolution. And the matched part wouldn't be highlighted as it is now. I think the actual matched text can help a lot with giving context. Also: the slide preview image is of no use to a blind person, so accessibility-wise it is useful to show the matched text. But of course, showing gibberish is not useful either. So I will make it configurable at least.
I'm strongly in favour of keeping the text extract also when there is a slide. Accessibility-wise and also for all other users the shown text can help with giving context.

> 
> > * For transcript results, maybe add quotation marks? "…Entwicklung scheint das Zauberwort der Zukun 211 sein und… "
> 
> That's a good idea. Unfortunately, there is hidden complexity. There are many different kinds of quotation marks and using the right one depends on the language of the text. „German”, “English”, »German alternative and Danish«, « French », «Swiss French and German?» ... just to name a few of languages/cultures familiar to us. [There are way more](https://en.wikipedia.org/wiki/Quotation_mark#Summary_table). And the problem is knowing the language of the search result. We can check the language of the video (which is already annoying but ok), but that might not be set or not match the text in question. It might be possible to only use one style of quotation mark by having them far away from the text and stylized somehow, but still... I would probably just leave as is.
I'm fine to leave it as is.

> 
> > * Maybe distinguish the two in color in the timeline?
> > * Or are you doing this already? Search "video" and see David's talk:
> 
> Yes, this is already done. It's just a different brightness of the color, not a different hue. But we don't really have a different-hue-color we could use (except the "danger color", but that's not fitting).
I don't get this point. Lets discuss it in our todays meeting

**Comment by dagraf** [comment]
> > > * The size of the white canvas in which slide and transcript results are presented seems to depend on the lenght of the text found which makes it uneasy to look at (first result for "video", Matthias' talk).
> > 
> > 
> > Yeah true, fair point, will try to fix that.
> 
> I spoke too quickly: how would you even fix this? The problematic case is when the matched text is long. I think there are only these options, none of which is optimal:
> 
>     * Cut off the matched text, only showing a very tiny section around the match: I think the length of the shown text is already very short and reducing the context around the match isn't great.
> 
>     * Let the text wrap into more than two lines: also not sure how useful that is.
> 
>     * Make the image larger: blowing up these low resolution images isn't great, and having such a big tooltip isn't great either.
> 
> 
> I guess I can somewhat improve the situation by making the image a bit larger still and reducing the max-width of the tooltip a bit (i.e. cutting more text off). But I don't think I can come up with a perfect solution.

I would suggest to fix the width of the shown white canvas (see screenshot for a possible size) and cut off the matched text so it always fits into this canvas and does NOT warp into more than one line.
Bildschirmfoto 2024-09-04 um 14 54 40

I would suggest to

**Comment by oas777** [comment]
PS: No one mentioned the color of the underscore, right? 
grafik
Where does that come from in a blueish color scheme?

**Comment by LukasKalbertodt** [comment]
Okay, I tried to change what we discussed. However, I decided to have different "tooltip size" behavior depending on whether a preview image is available:

- No image: the tooltip is sized dynamically according to the amount of text. 
- With image: the tooltip has a fixed height, the width can vary just a tiny bit, if there is a lot of text.

The fixed height and some other changes make hovering over parts of the timeline a lot smoother I think, i.e. less jumping around of the tooltip. Let me know what you think about the remaining width-flexibility for large text. 

---

As a second step, I added a faint icon to the lower bottom corner of the tooltip to show what kind of result we are talking about. Please also let me know what you think.

**Comment by oas777** [comment]
Thanks, Lukas. The icon are very helpful, though the one for slide content could be more along the lines of
.
"Tooltip size" is better/more coherent also, though there is a (different) problem with tooltips below the timeline:
grafik

**Comment by LukasKalbertodt** [comment]
> there is a (different) problem with tooltips below the timeline

Good catch, should be fixed now.

>  the one for slide content could be more along the lines of

Is that a Lucide icon you found there? I could not find it anywhere. I checked Lucide and found these that somewhat match your suggestion:  [`scan-text`](https://lucide.dev/icons/scan-text), [`letter-text`](https://lucide.dev/icons/letter-text) and [`file-text`](https://lucide.dev/icons/file-text).

**Comment by oas777** [comment]
No, that was me searching Google. I like "letter-text".

**Comment by oas777** [comment]
Looks good, thanks. Allow me to mention that two tooltips are visible at the same time sometimes when you hover the timeline (slowly  and) horizontally:
grafik
Couldn't detect a logic for this happening and maybe it's only due to the number of results.

**Comment by LukasKalbertodt** [comment]
> two tooltips are visible at the same time sometimes

Yes I noticed this too :/ I think it's a glitch in some other package and we can't do a whole lot about it. But we can still look into it again.

We are currently still trying to fix some weird technical bug. But once that's done, this can be merged I think.

**Comment by owi92** [comment]
Nice, so apart from the glitch Lukas and Olaf mentioned above everybody's happy and the technical bug also appears to be fixed. I think this is a great improvement. I'll merge this now 👍

**Pull request merged**